### PR TITLE
add initializers and BaseModule for unified parameter initialization

### PR DESCRIPTION
**Pull request by MeowZheng**
I mainly revised 3 files:
1. In weight_init.py, add ``BiasInitWithProb``, ``ConstantInit``, ``KaimingInit``,`` NormalInit``, ``PretrainedInit``, ``UniformInit``, ``XavierInit`` classes, and register them in ``"INITIALIZERS"`` registry; add ``initialize`` function to initialize parameters with ``"init_cfg"``.
2. In checkpoint.py, add ``_load_checkpoint_with_prefix`` function.
3. Add base_module.py,  add ``"BaseModule"`` and only implement ``init_weight`` func for parameters initialization 

## Design
Model intialization in OpenMMLab uses ``init_cfg``, ``BaseModule::init_weight``, ``initialize``, and ``INITIALIZERS`` registry together. Users can initialize their models with following two steps:
1. Define ``init_cfg`` for a model or its components in ``model_cfg``, but ``init_cfg`` of children components have higher priority and will override ``init_cfg`` of parents modules.
2. Build model as usual, but call ``model.init_weight()`` method explicitly, and model parameters will be initialized as configuration.

The high-level workflow of initialization in OpenMMLab is:
model_cfg(init_cfg) -> build_from_cfg -> model -> init_weight() -> initialize(self, self.init_cfg) -> children's init_weight()

### init_cfg 
it is a dict or vf[dict], and contains:
+ `type` - a str containing the initializer in `INTIALIZERS`, and it must be assigned a valid value if init_cfg is used. It is followed by arguments of the initializer if default values cannot meet the demand.
+ `layer` - a str or [str] containing the names of layers with learnable parameters that will be initialized.
+ `override` - a dict or [dict] containing the sub-modules that do not inherit from BaseModule and whose initialization configuration is different from other layers' which are in 'layers'. Initializer defined in `type` will work for all layers defined in `layer`, so if sub-modules are not derived Classes of `BaseModule` but can be initialized as same ways of layers in `layer`, it does not need to use `override`. `override` contains:
    + `type` followed by arguments of initializer;
    + `name` to indicate sub-module which will be initialized.

### BaseModule
`BaseModule` is the base module for all modules in OpenMMLab. `init_weight` method of `BaseModule` can initialize itself parameters using `initialize(module, init_cfg)` function in mmcv, and call sub-components' `init_weight()`method.

### initialize(module, init_cfg)
+ `module` - the module will be initialized.
+ `init_cfg` -  initialization configuration dict.

### INITIALIZERS registry
OpenMMLab has implemented 7 initializers including ``Constant``, ``Xavier``, ``Normal``, ``Uniform``,  ``Kaiming``, ``Pretrained`` and ``BiasProb``(only for  bias initialization), and registers them in `INITIALIZERS`

Taking advantage of "buider&registry" mechanism of OpenMMLab, ``INITIALIZERS`` can be easily extended by implementing new initializer classes and registering them in ``INITIALIZERS``.

## Usages
users initialize models of OpenMMLab, just need two steps: 1. define init_cfg; 2. build model and call ``model.init_weight()``.

### define init_cfg for model
``FooModel``, ``FooConv1d``, ``FooConv2d`` and ``FooLinear`` are deritived classes from ``BaseModule``. If we would like to initialize all weight of linear layer as 1 and bias as 2, all weight of conv1d layer as 3 and bias as 4, all weight of conv2d layer as 5 and bias as 6 of `FooModel`, we can define model_cfg and init_cfg as following 
```python
model_cfg = dict(
    type="FooModel",
    init_cfg=[
        dict(type='Constant', val=1, bias=2, layer='Linear'),
        dict(type='Constant', val=3, bias=4, layer='Conv1d'),
        dict(type='Constant', val=5, bias=6, layer='Conv2d')
    ],
    component1=dict(type='FooConv1d'),
    component2=dict(type='FooConv2d'),
    component3=dict(type='FooLinear'),
    component4=dict(
        type='FooLinearConv1d',
        linear=dict(type='FooLinear'),
        conv1d=dict(type='FooConv1d')))
```
After this, we build a FooModel instance and call init_weight  
```python
model = build_from_cfg(model_cfg, FOOMODELS)
model.init_weight()
```

### define init_cfg nestedly
init_cfg of sub-modules will override the parents', like:
```python
model_cfg = dict(
    type="FooModel",
    init_cfg=[
        dict(type='Constant', val=1, bias=2,layer='Linear',
            override=dict(type='Constant', name='reg', val=13, bias=14)),
        dict(type='Constant', val=3, bias=4, layer='Conv1d'),
        dict(type='Constant', val=5, bias=6, layer='Conv2d'),
    ],
    component1=dict(
        type='FooConv1d', init_cfg=dict(type='Constant', val=7, bias=8)),
    component2=dict(
        type='FooConv2d', init_cfg=dict(type='Constant', val=9, bias=10)),
    component3=dict(type='FooLinear'),
    component4=dict(
        type='FooLinearConv1d',
        linear=dict(type='FooLinear'),
        conv1d=dict(type='FooConv1d')))
```
after ``model = build_from_cfg(model_cfg, FOOMODELS)`` and ``model.init_weight()``, parameters will be 
model (FooModel)
+ component1 (FooConv1d, weight=7, bias=8)
+ component2 (FooConv2d, weight=9, bias=10)
+ component3 (FooLinear, weight=1, bias=2)
+ component4 (FooLinearConv1d)
    + linear (FooLinear, weight=1, bias=2)
    + conv1d (FooConv1d, weight=3, bias=4)
+ reg (nn.Linear, weight=13, bias=14)

## Migration
1. If models inherit from ``nn.Module``, must inherit from ``BaseModule``.
2. Add `init_cfg` argument in `__init__` of derived classes, and set default value for `init_cfg`:
If `init_weight` in current classes is recursively call `init_weight` of childrend's modules, such as
```python
def init_weights(self, pretrained):
    """Initialize the weights in head.

    Args:
        pretrained (str, optional): Path to pre-trained weights.
            Defaults to None.
    """
    if self.with_shared_head:
        self.shared_head.init_weights(pretrained=pretrained)
    if self.with_bbox:
        self.bbox_roi_extractor.init_weights()
        self.bbox_head.init_weights()
    if self.with_mask:
        self.mask_head.init_weights()
        if not self.share_roi_extractor:
            self.mask_roi_extractor.init_weights()
```
just set ``init_cfg = None``. Otherwise, set ``init_cfg`` value according to current code in `init_weight`, e.g.
```python
# init_weight from retina_head
def init_weights(self):
    """Initialize weights of the head."""
    for m in self.cls_convs:
        normal_init(m.conv, std=0.01)
    for m in self.reg_convs:
        normal_init(m.conv, std=0.01)
    bias_cls = bias_init_with_prob(0.01)
    normal_init(self.retina_cls, std=0.01, bias=bias_cls)
    normal_init(self.retina_reg, std=0.01)
```
the init_cfg must be
```python
init_cfg = dict(
    type='Normal',
    layers='Conv2d',
    std=0.01,
    override=dict(type='Normal', name='retina_cls',std=0.01,
                bias=dict(type='BiasProb', prior_prob=0.01)))
```
3. Backward compatibility for `pretrained` in previous config file. Add following code in ``__init__`` of derived classes from ``BaseModule``
```python
if pretrained is not None:
    warnings.warn('DeprecationWarning: pretrained is a deprecated \
        key, please consider using init_cfg')
    self.init_cfg = dict(type='pretrained', checkpoint=pretrained)
``` 
4. There is no need to override `init_weight` method in derived classes, if users call ``init_weight`` of sub-components, or call ``init_weight`` of model twice, there will be a warning "This module has been initialized, please call initialize(module, init_cfg) to reinitialize it".
5. Call ``model.init_weight()``, after building models. Please pay attention to it, as this is an additional action for models in OpenMMLab.  

## BC-breaking
Please inform users to call ``model.init_weight()`` after building models in tutorals.

### CODE REVIEW

**Comment by ZwwWayne** [comment]
May add design, usages, migration, and BC-breakings in the PR messages and documentation for both discussion and reference for users.

File: mmcv/runner/base_module.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# Copyright (c) Open-MMLab. All rights reserved.
+import warnings
+from abc import ABCMeta
+
+import torch.nn as nn
+
+from mmcv.cnn import initialize
+
+
+class BaseModule(nn.Module, metaclass=ABCMeta):
```
Comment (ZwwWayne) [inline mmcv/runner/base_module.py:10]: Should this be put into mmcv.cnn.bricks?

**Comment by ZwwWayne** [comment]
The contents of PR messages should also be put into the tutorial to serve as documentation.

**Comment by ZwwWayne** [comment]
Please resolve conflicts as the load_checkpoint has been refactored.

File: mmcv/cnn/utils/weight_init.py [hunk 2]
```diff
@@ -57,10 +65,399 @@ def caffe2_xavier_init(module, bias=0):
...
+    distribution :math:`\mathcal{N}(\text{mean}, \text{std}^2)`.
+
+    Args:
+        mean (int | float):the mean of the normal distribution.
+        Defaults to 0.
+        std (int | float): the standard deviation of the normal
+        distribution. Defaults to 1.
+        bias (int | float | dict): the value to fill the bias or
+        define initialization type for bias. Defaults to 0.
+        layers (str | [str], optional): the layer will be initialized.
```
Comment (ZwwWayne) [inline mmcv/cnn/utils/weight_init.py:182]: [str] -> list[str], also fix similar issues in other places.

File: mmcv/cnn/utils/weight_init.py [hunk 3]
```diff
@@ -57,10 +65,399 @@ def caffe2_xavier_init(module, bias=0):
...
+            else:
+                raise TypeError(
+                    f'layers must be str or [str], but gor {type(self.layers)}'
+                )
+
+
+@INITIALIZERS.register_module(name='Kaiming')
+class KaimingInit(BaseInit):
+    r"""Initialize module paramters with the valuse according to the method
+    described in `Delving deep into rectifiers: Surpassing human-level
```
Comment (ZwwWayne) [inline mmcv/cnn/utils/weight_init.py:254]: May add paper link?

File: mmcv/cnn/utils/weight_init.py [hunk 4]
```diff
@@ -57,10 +65,399 @@ def caffe2_xavier_init(module, bias=0):
...
+
+
+@INITIALIZERS.register_module(name='Uniform')
+class UniformInit(BaseInit):
+    r"""Initialize module parameters with values drawn from the uniform
+    distribution :math:`\mathcal{U}(a, b)`.
+
+    Args:
+        a (int | float): the lower bound of the uniform
+        distribution. Defaults to 0.
```
Comment (ZwwWayne) [inline mmcv/cnn/utils/weight_init.py:218]: May check the doc in the local host to see whether the results are expected. Usually we will leave an indent when switching lines for an argument

File: tests/test_runner/test_checkpoint.py [hunk 5]
```diff
@@ -152,10 +155,44 @@ def test_load_pavimodel_dist():
...
+        torch.save(model.state_dict(), 'model.pth')
+        prefix = 'conv2d'
+        state_dict = _load_checkpoint_with_prefix(prefix, 'model.pth')
+        assert torch.equal(model.conv2d.state_dict()['weight'],
+                           state_dict['weight'])
+        assert torch.equal(model.conv2d.state_dict()['bias'],
+                           state_dict['bias'])
+
+        prefix = 'back'
+        with pytest.raises(AssertionError):
```
Comment (ZwwWayne) [inline tests/test_runner/test_checkpoint.py:186]: Add comments to indicate what error you are checking.

File: mmcv/cnn/resnet.py [hunk 6]
```diff
@@ -266,6 +265,7 @@ def __init__(self,
     def init_weights(self, pretrained=None):
         if isinstance(pretrained, str):
             logger = logging.getLogger()
+            from ..runner import load_checkpoint
```
Comment (ZwwWayne) [inline mmcv/cnn/resnet.py:268]: Shall we start to use BaseModule for resnet and vgg as examples?

File: mmcv/cnn/utils/weight_init.py [hunk 7]
```diff
@@ -57,10 +65,399 @@ def caffe2_xavier_init(module, bias=0):
...
+
+
+def _initialize_cases(module, cases):
+    if isinstance(cases, list):
+        for case in cases:
+            name = case.pop('name', None)
+            if hasattr(module, name):
+                _initialize(getattr(module, name), case)
+            else:
+                raise RuntimeError(f'module did not have attribute {name}')
```
Comment (ZwwWayne) [inline mmcv/cnn/utils/weight_init.py:376]: May check those errors in unit tests.

File: mmcv/cnn/resnet.py (hunk 6 above)
Comment (hellock) [reply mmcv/cnn/resnet.py:268]: Maybe in a separated PR?

**Comment by ZwwWayne** [comment]
LGTM now.

**Review by hellock — APPROVED** [review]

File: mmcv/cnn/utils/weight_init.py [hunk 8]
```diff
@@ -57,10 +65,367 @@ def caffe2_xavier_init(module, bias=0):
...
+        module (``torch.nn.Module``): the module will be initialized.
+        init_cfg (dict | list[dict]): initialization configuration dict to
+            define initializer. OpenMMLab has implemented 7 initializers
+            including ``Constant``, ``Xavier``, ``Normal``, ``Uniform``,
+            ``Kaiming``, ``Pretrained`` and ``BiasProb`` for bias
+            initialization.
+
+    Example:
+        >>> module = nn.Linear(2, 3, bias=True)
+        >>> init_cfg = dict(type='Constant', val =1 , bias =2)
```
Comment (dreamerlin) [inline mmcv/cnn/utils/weight_init.py:376]: redunant blank

File: mmcv/cnn/utils/weight_init.py [hunk 9]
```diff
@@ -57,10 +65,367 @@ def caffe2_xavier_init(module, bias=0):
...
+
+def _initialize_override(module, override):
+    if not isinstance(override, (dict, list)):
+        raise TypeError(
+            f'override must be a dict or list, but got {type(override)}')
+
+    override = [override] if isinstance(override, dict) else override
+
+    for override_ in override:
+        name = override_.pop('name', None)
```
Comment (dreamerlin) [inline mmcv/cnn/utils/weight_init.py:356]: None may lead to error `TypeError: hasattr(): attribute name must be string`

File: mmcv/runner/base_module.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        # if pretrained is not None:
+        #     warnings.warn('DeprecationWarning: pretrained is a deprecated \
+        #         key, please consider using init_cfg')
+        #     self.init_cfg = dict(type='Pretrained', checkpoint=pretrained)
+
+    @property
+    def is_init(self):
+        return self._is_init
+
+    def init_weight(self):
```
Comment (dreamerlin) [inline mmcv/runner/base_module.py:38]: `init_weight` or `init_weights` ?

File: mmcv/cnn/utils/weight_init.py (hunk 9 above)
Comment (MeowZheng) [reply mmcv/cnn/utils/weight_init.py:356]: [#930](https://github.com/open-mmlab/mmcv/pull/930)

File: mmcv/cnn/utils/weight_init.py (hunk 8 above)
Comment (MeowZheng) [reply mmcv/cnn/utils/weight_init.py:376]: #930

**Comment by apanand14** [comment]
File "tools/train.py", line 163, in main
    model.init_weights()
  File "C:\Users\topseven\anaconda3\envs\mmcv\lib\site-packages\mmcv\runner\base_module.py", line 117, in init_weights
    m.init_weights()
**TypeError: init_weights() missing 1 required positional argument: 'pretrained'**

**Pull request merged**